Source/DafnyCore/AST/TypeUtil.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+
+namespace Microsoft.Dafny;
+
+public class TypeUtil {
+  public static Dictionary<TypeParameter, Type> TypeSubstitutionMap(List<TypeParameter> formals, List<Type> actuals) {
+    Contract.Requires(formals != null);
+    Contract.Requires(actuals != null);
+    Contract.Requires(formals.Count == actuals.Count);
+    var subst = new Dictionary<TypeParameter, Type>();
+    for (int i = 0; i < formals.Count; i++) {
+      subst.Add(formals[i], actuals[i]);
+    }
+    return subst;
+  }
+
+  /// <summary>
+  /// If the substitution has no effect, the return value is pointer-equal to 'type'
+  /// </summary>
+  public static Type SubstType(Type type, Dictionary<TypeParameter, Type> subst) {
```

Comment:
Consider making this an extension method, so you can write `someType.Substitute(substitutionMap)`

Even better would be to have this be an abstract method defined on `Type`, so the code is more local and there's a static check against forgetting to add cases for this behavior.